torch/utils/benchmark/utils/timer.py

```diff
@@ -21,6 +22,57 @@ def timer() -> float:
     timer = timeit.default_timer
 
 
+PYTHON = "Python"
+CPP = "C++"
+
+
+class CPPTimer:
+    def __init__(
+        self,
+        stmt: str,
+        setup: str,
+        timer: Callable[[], float],
+        globals: Dict[str, Any],
+    ) -> None:
+        if timer is not timeit.default_timer:
+            # FIXME: We should be able to automatically detect GPU operations.
+            raise ValueError(
+                "When timing C++, timer must be `timeit.default_timer`. If you "
+                "did not specify `timer=...`, this means that a timer with "
+                "CUDA syncronization was chosen, and you should pass "
+                "`timer=timeit.default_timer` to the Timer's constructor to "
+                "declare that CUDA syncronization is not needed."
+            )
+
+        if len(globals) > 1 or "torch" not in globals:
+            # Timer will add `torch` as a convenience.
+            raise ValueError("C++ timing does not support globals.")
+
+        if globals.get("torch", torch) is not torch:
+            # cpp_extension is going to link against `libtorch.so` that comes
+            # from `import torch`, so changing `torch` in globals would be
+            # silently ignored.
+            raise ValueError(
+                "Swapping out `torch` is not allowed for C++ snippets."
+            )
+
+        self._stmt: str = textwrap.dedent(stmt)
+        self._setup: str = textwrap.dedent(setup)
+        self._timeit_module: Optional[cpp_jit.TimeitModuleType] = None
+
+    def timeit(self, number: int) -> float:
+        if self._timeit_module is None:
+            self._timeit_module = cast(
```

Comment:
Hmm, how come you need a cast here...

Reply:
The type checker can't deduce the type. I've added a comment explaining why.

Reply:
A better strategy, when performance doesn't matter (which I don't think it does here), is instead do `assert isinstance(r, cpp_jit.TimeitModuleType)`. This has the added benefit of actually calling you out at runtime when you get it wrong. I much prefer Haskell's naming for its equivalent function: `unsafeCoerce`